### PR TITLE
Keymanager APIs: fee recipient api

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Feature

related to #10292
**What does this PR do? Why is it needed?**

This PR adds the get, post, and delete APIs as specified by the [keymanager APIs](https://github.com/ethereum/keymanager-APIs/blob/master/apis/fee_recipient.yaml). This will allow users to update a fee recipient for a given validator public key. This api activated in prysm via the `--web` command and works similar to other keymanager APIs. 

note: the fee recipient is updated in the validator client but will update on the beacon chain/builder only during the set intervals ( when the validator client turns on, when a validator key becomes active, or at the beginning of an epoch)

### CODE REVIEW

File: validator/rpc/standard_api.go [hunk 1]
```diff
@@ -382,3 +388,117 @@ func groupDeleteRemoteKeysErrors(req *ethpbservice.DeleteRemoteKeysRequest, erro
 	}
 	return statuses
 }
+
+// ListFeeRecipientByPubkey returns the public key to eth address mapping object to the end user.
+func (s *Server) ListFeeRecipientByPubkey(ctx context.Context, req *ethpbservice.ByPubkeyRequest) (*ethpbservice.GetFeeRecipientByPubkeyResponse, error) {
+	if s.validatorService == nil {
+		return nil, status.Error(codes.FailedPrecondition, "Validator service not ready.")
```
Comment (rauljordan) [inline validator/rpc/standard_api.go:395]:
```suggestion
		return nil, status.Error(codes.FailedPrecondition, "Validator service not ready")
```

File: validator/rpc/standard_api.go [hunk 2]
```diff
@@ -382,3 +388,117 @@ func groupDeleteRemoteKeysErrors(req *ethpbservice.DeleteRemoteKeysRequest, erro
...
+	}
+	validatorKey := req.Pubkey
+	if len(validatorKey) != fieldparams.BLSPubkeyLength {
+		return nil, status.Errorf(
+			codes.InvalidArgument, "provided public key in path is not a valid bls public key, please check for correct hex format starting with 0x")
```
Comment (rauljordan) [inline validator/rpc/standard_api.go:400]: Technically, it should only be 0x prefixed hex if this is called via the API middleware. If called via grpc, it needs to be raw bytes. However, no one will be calling this through gRPC I guess, unless it is done programmatically. Just to be safe, I would amend the error message to say the key needs to have the right length and and printf the field params value instead of telling the user it needs to be a 0x prefixed hex string

File: validator/rpc/standard_api.go [hunk 3]
```diff
@@ -382,3 +388,117 @@ func groupDeleteRemoteKeysErrors(req *ethpbservice.DeleteRemoteKeysRequest, erro
...
+	}
+	defaultOption := validatorServiceConfig.DefaultProposerOption()
+	encoded := hexutil.Encode(req.Ethaddress)
+	if !common.IsHexAddress(encoded) {
+		return nil, status.Errorf(
+			codes.InvalidArgument, "fee recipient is not a valid eth1 address")
```
Comment (rauljordan) [inline validator/rpc/standard_api.go:448]:
```suggestion
			codes.InvalidArgument, "fee recipient is not a valid Ethereum address")
```

File: validator/rpc/standard_api.go [hunk 4]
```diff
@@ -382,3 +388,117 @@ func groupDeleteRemoteKeysErrors(req *ethpbservice.DeleteRemoteKeysRequest, erro
...
+// DeleteFeeRecipientByPubkey updates the eth address mapped to the public key to the default fee recipient listed
+func (s *Server) DeleteFeeRecipientByPubkey(ctx context.Context, req *ethpbservice.ByPubkeyRequest) (*empty.Empty, error) {
+	if s.validatorService == nil {
+		return nil, status.Error(codes.FailedPrecondition, "Validator service not ready.")
+	}
+	validatorKey := req.Pubkey
+	if len(validatorKey) != fieldparams.BLSPubkeyLength {
+		return nil, status.Errorf(
+			codes.InvalidArgument, "provided public key in path is not a valid bls public key, please check for correct hex format starting with 0x")
+	}
```
Comment (rauljordan) [inline validator/rpc/standard_api.go:488]: I think this is duplicated 3 times. Might as well extract into a func

File: api/gateway/apimiddleware/api_middleware.go [hunk 5]
```diff
@@ -111,7 +111,7 @@ func (m *ApiProxyMiddleware) WithMiddleware(path string) http.HandlerFunc {
 			}
 		}
 
-		if req.Method == "DELETE" {
+		if req.Method == "DELETE" && req.Body != http.NoBody {
```
Comment (rkapka) [inline api/gateway/apimiddleware/api_middleware.go:114]: Why this change?

File: config/validator/service/proposer-settings.go [hunk 6]
```diff
@@ -33,3 +34,11 @@ type ProposerOption struct {
 	FeeRecipient common.Address
 	GasLimit     uint64
 }
+
+//DefaultProposerOption returns a Proposer Option with defaults filled
```
Comment (rkapka) [inline config/validator/service/proposer-settings.go:38]:
```suggestion
// DefaultProposerOption returns a Proposer Option with defaults filled
```

File: proto/eth/service/key_management.proto [hunk 7]
```diff
@@ -206,3 +268,20 @@ message DeletedRemoteKeysStatus {
   string message = 2;
 }
 
+
+message ByPubkeyRequest {
```
Comment (rkapka) [inline proto/eth/service/key_management.proto:272]: `ByPubkey` sounds incomplete. Maybe just `PubkeyRequest`?

File: validator/rpc/standard_api.go [hunk 8]
```diff
@@ -382,3 +388,117 @@ func groupDeleteRemoteKeysErrors(req *ethpbservice.DeleteRemoteKeysRequest, erro
...
+	if s.validatorService.ProposerSettings.DefaultConfig != nil {
+		defaultFeeRecipient = s.validatorService.ProposerSettings.DefaultConfig.FeeRecipient.Bytes()
+	}
+	return &ethpbservice.GetFeeRecipientByPubkeyResponse{
+		Data: &ethpbservice.GetFeeRecipientByPubkeyResponse_FeeRecipient{
+			Pubkey:     validatorKey,
+			Ethaddress: defaultFeeRecipient,
+		},
+	}, nil
+
```
Comment (rkapka) [inline validator/rpc/standard_api.go:430]:
```suggestion
```

File: validator/rpc/standard_api.go [hunk 9]
```diff
@@ -382,3 +388,117 @@ func groupDeleteRemoteKeysErrors(req *ethpbservice.DeleteRemoteKeysRequest, erro
...
+// ListFeeRecipientByPubkey returns the public key to eth address mapping object to the end user.
+func (s *Server) ListFeeRecipientByPubkey(ctx context.Context, req *ethpbservice.ByPubkeyRequest) (*ethpbservice.GetFeeRecipientByPubkeyResponse, error) {
+	if s.validatorService == nil {
+		return nil, status.Error(codes.FailedPrecondition, "Validator service not ready")
+	}
+	validatorKey := req.Pubkey
+	if len(validatorKey) != fieldparams.BLSPubkeyLength {
+		return nil, status.Errorf(
+			codes.InvalidArgument, "provided public key in path is not a valid bls public key, please check for correct hex format starting with 0x")
```
Comment (rkapka) [inline validator/rpc/standard_api.go]:
```suggestion
			codes.InvalidArgument, "Provided public key in path is not a valid bls public key, please check for correct hex format starting with 0x")
```

File: validator/rpc/standard_api.go [hunk 10]
```diff
@@ -382,3 +388,117 @@ func groupDeleteRemoteKeysErrors(req *ethpbservice.DeleteRemoteKeysRequest, erro
...
+
+// SetFeeRecipientByPubkey updates the eth address mapped to the public key.
+func (s *Server) SetFeeRecipientByPubkey(ctx context.Context, req *ethpbservice.SetFeeRecipientByPubkeyRequest) (*empty.Empty, error) {
+	if s.validatorService == nil {
+		return nil, status.Error(codes.FailedPrecondition, "Validator service not ready.")
+	}
+	validatorKey := req.Pubkey
+	if len(validatorKey) != fieldparams.BLSPubkeyLength {
+		return nil, status.Errorf(
+			codes.InvalidArgument, "provided public key in path is not a valid bls public key, please check for correct hex format starting with 0x")
```
Comment (rkapka) [inline validator/rpc/standard_api.go]:
```suggestion
			codes.InvalidArgument, "Provided public key in path is not a valid bls public key, please check for correct hex format starting with 0x")
```

File: validator/rpc/standard_api.go [hunk 11]
```diff
@@ -382,3 +388,117 @@ func groupDeleteRemoteKeysErrors(req *ethpbservice.DeleteRemoteKeysRequest, erro
...
+	}
+	defaultOption := validatorServiceConfig.DefaultProposerOption()
+	encoded := hexutil.Encode(req.Ethaddress)
+	if !common.IsHexAddress(encoded) {
+		return nil, status.Errorf(
+			codes.InvalidArgument, "fee recipient is not a valid Ethereum address")
```
Comment (rkapka) [inline validator/rpc/standard_api.go:446]:
```suggestion
			codes.InvalidArgument, "Fee recipient is not a valid Ethereum address")
```

File: validator/rpc/standard_api.go [hunk 12]
```diff
@@ -382,3 +388,117 @@ func groupDeleteRemoteKeysErrors(req *ethpbservice.DeleteRemoteKeysRequest, erro
...
+
+// DeleteFeeRecipientByPubkey updates the eth address mapped to the public key to the default fee recipient listed
+func (s *Server) DeleteFeeRecipientByPubkey(ctx context.Context, req *ethpbservice.ByPubkeyRequest) (*empty.Empty, error) {
+	if s.validatorService == nil {
+		return nil, status.Error(codes.FailedPrecondition, "Validator service not ready.")
+	}
+	validatorKey := req.Pubkey
+	if len(validatorKey) != fieldparams.BLSPubkeyLength {
+		return nil, status.Errorf(
+			codes.InvalidArgument, "provided public key in path is not a valid bls public key, please check for correct hex format starting with 0x")
```
Comment (rkapka) [inline validator/rpc/standard_api.go]:
```suggestion
			codes.InvalidArgument, "Provided public key in path is not a valid bls public key, please check for correct hex format starting with 0x")
```

File: api/gateway/apimiddleware/process_field.go [hunk 13]
```diff
@@ -100,6 +101,20 @@ func base64ToHexProcessor(v reflect.Value) error {
 	return nil
 }
 
+func base64ToChecksumAddressProcessor(v reflect.Value) error {
```
Comment (rkapka) [inline api/gateway/apimiddleware/process_field.go:104]: Please add unit tests

File: validator/rpc/standard_api.go [hunk 14]
```diff
@@ -382,3 +388,117 @@ func groupDeleteRemoteKeysErrors(req *ethpbservice.DeleteRemoteKeysRequest, erro
...
+		if found {
+			proposerOption.FeeRecipient = common.BytesToAddress(req.Ethaddress)
+		} else {
+			s.validatorService.ProposerSettings.ProposeConfig[bytesutil.ToBytes48(validatorKey)] = &pOption
+		}
+	}
+	// override the 200 success with 202 according to the specs
+	if err := grpc.SetHeader(ctx, metadata.Pairs("x-http-code", "202")); err != nil {
+		return &empty.Empty{}, fmt.Errorf("could not set custom success code header: %w", err)
+	}
```
Comment (rkapka) [inline validator/rpc/standard_api.go:473]: I don't consider this an error condition. But if you do want to return an error here, then you should use `status.Errorf` and start with an uppercase.

File: validator/rpc/standard_api.go [hunk 15]
```diff
@@ -382,3 +388,117 @@ func groupDeleteRemoteKeysErrors(req *ethpbservice.DeleteRemoteKeysRequest, erro
...
+	if s.validatorService.ProposerSettings != nil && s.validatorService.ProposerSettings.ProposeConfig != nil {
+		proposerOption, found := s.validatorService.ProposerSettings.ProposeConfig[bytesutil.ToBytes48(validatorKey)]
+		if found {
+			proposerOption.FeeRecipient = defaultFeeRecipient
+		}
+	}
+	// override the 200 success with 204 according to the specs
+	if err := grpc.SetHeader(ctx, metadata.Pairs("x-http-code", "204")); err != nil {
+		return &empty.Empty{}, fmt.Errorf("could not set custom success code header: %w", err)
+	}
```
Comment (rkapka) [inline validator/rpc/standard_api.go:499]: Same comment as above.

File: api/gateway/apimiddleware/api_middleware.go (hunk 5 above)
Comment (james-prysm) [reply api/gateway/apimiddleware/api_middleware.go:114]: my delete function has no body sent, if i don't have this statement then it tries to marshal a body that doesn't exist into another entity that doesn't exist.

File: validator/rpc/standard_api.go (hunk 9 above)
Comment (james-prysm) [reply validator/rpc/standard_api.go]: using a variable now

File: api/gateway/apimiddleware/process_field.go (hunk 13 above)
Comment (james-prysm) [reply api/gateway/apimiddleware/process_field.go:104]: added to process_request_test.go

File: validator/rpc/standard_api.go [hunk 16]
```diff
@@ -382,3 +388,121 @@ func groupDeleteRemoteKeysErrors(req *ethpbservice.DeleteRemoteKeysRequest, erro
...
+			Pubkey:     validatorKey,
+			Ethaddress: defaultFeeRecipient,
+		},
+	}, nil
+}
+
+// SetFeeRecipientByPubkey updates the eth address mapped to the public key.
+func (s *Server) SetFeeRecipientByPubkey(ctx context.Context, req *ethpbservice.SetFeeRecipientByPubkeyRequest) (*empty.Empty, error) {
+	if s.validatorService == nil {
+		return nil, status.Error(codes.FailedPrecondition, "Validator service not ready.")
```
Comment (rkapka) [inline validator/rpc/standard_api.go:435]:
```suggestion
		return nil, status.Error(codes.FailedPrecondition, "Validator service not ready")
```

File: validator/rpc/standard_api.go [hunk 17]
```diff
@@ -382,3 +388,121 @@ func groupDeleteRemoteKeysErrors(req *ethpbservice.DeleteRemoteKeysRequest, erro
...
+	if err := grpc.SetHeader(ctx, metadata.Pairs("x-http-code", "202")); err != nil {
+		return &empty.Empty{}, status.Errorf(codes.Internal, "Could not set custom success code header: %v", err)
+	}
+	return &empty.Empty{}, nil
+}
+
+// DeleteFeeRecipientByPubkey updates the eth address mapped to the public key to the default fee recipient listed
+func (s *Server) DeleteFeeRecipientByPubkey(ctx context.Context, req *ethpbservice.PubkeyRequest) (*empty.Empty, error) {
+	if s.validatorService == nil {
+		return nil, status.Error(codes.FailedPrecondition, "Validator service not ready.")
```
Comment (rkapka) [inline validator/rpc/standard_api.go:479]:
```suggestion
		return nil, status.Error(codes.FailedPrecondition, "Validator service not ready")
```

**Review by rkapka — APPROVED** [review]
